vespajlib/src/main/java/com/yahoo/slime/Value.java

```diff
@@ -84,4 +85,82 @@ public final String toString() {
             return "null";
         }
     }
+
+    private static class Equal {
+        protected final Inspector rhsInspector;
+
+        protected boolean equal = true;
+
+        public Equal(Inspector rhsInspector) { this.rhsInspector = rhsInspector; }
+
+        public boolean isEqual() { return equal; }
+    }
+
+    private static class EqualArray extends Equal implements ArrayTraverser {
+        public EqualArray(Inspector rhsInspector) { super(rhsInspector); }
+
+        @Override
+        public void entry(int idx, Inspector inspector) {
+            if (equal) {
+                equal = inspector.equalTo(rhsInspector.entry(idx));
+            }
+        }
+    }
+
+    private static class EqualObject extends Equal implements ObjectTraverser {
+        public EqualObject(Inspector rhsInspector) { super(rhsInspector); }
+
+        @Override
+        public void field(String name, Inspector inspector) {
+            if (equal) {
+                equal = inspector.equalTo(rhsInspector.field(name));
+            }
+        }
+    }
+
+    @Override
+    public boolean equalTo(Inspector that) {
+        boolean equal = type() == that.type();
+
+        if (equal) {
+            switch (type()) {
+                case NIX:
+                    break;
```

Comment:
consider checking if the validity of the nixes are the same (we should also update this in C++)